src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationReloadToken.cs

```diff
@@ -33,11 +33,60 @@ public class ConfigurationReloadToken : IChangeToken
         /// <param name="callback">The callback to invoke.</param>
         /// <param name="state">State to be passed into the callback.</param>
         /// <returns>The <see cref="CancellationToken"/> registration.</returns>
-        public IDisposable RegisterChangeCallback(Action<object?> callback, object? state) => _cts.Token.Register(callback, state);
+        public IDisposable RegisterChangeCallback(Action<object?> callback, object? state)
+        {
+#if NETCOREAPP || NETSTANDARD2_1
+            try
+            {
+                return _cts.Token.UnsafeRegister(callback, state);
+            }
+            catch (ObjectDisposedException)
+            {
+                // Reset the flag so that we can indicate to future callers that this wouldn't work.
+                ActiveChangeCallbacks = false;
+            }
+#else
+            // Don't capture the current ExecutionContext and its AsyncLocals onto the token registration causing them to live forever
+            bool restoreFlow = false;
+            if (!ExecutionContext.IsFlowSuppressed())
+            {
+                ExecutionContext.SuppressFlow();
+                restoreFlow = true;
+            }
+
+            try
+            {
+                return _cts.Token.Register(callback, state);
+            }
+            catch (ObjectDisposedException)
+            {
+                // Reset the flag so that we can indicate to future callers that this wouldn't work.
+                ActiveChangeCallbacks = false;
+            }
+            finally
+            {
+                // Restore the current ExecutionContext
+                if (restoreFlow)
+                {
+                    ExecutionContext.RestoreFlow();
+                }
+            }
+#endif
+            return NullDisposable.Instance;
+        }
 
         /// <summary>
         /// Used to trigger the change token when a reload occurs.
         /// </summary>
         public void OnReload() => _cts.Cancel();
+
+        private sealed class NullDisposable : IDisposable
```

Comment:
We can just reuse

https://github.com/dotnet/runtime/blob/f1c3302534bc8ef2ee5b1604d4dcfe79de27046f/src/libraries/Common/src/Extensions/EmptyDisposable.cs#L8-L19

here. No need to duplicate it.